torch/_subclasses/fake_tensor.py

```diff
@@ -1414,6 +1536,281 @@ def __exit__(self, a, b, c):
+
+        if output is unassigned:
+            output = self._dispatch_impl(func, types, args, kwargs)
+
+        return output
+
+    def _cache_key(
+        self,
+        func: OpOverload,
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> _DispatchCacheKey:
+        """
+        Create a cache key given the dispatch args. Raises _BypassDispatchCache
+        for any situation that precludes caching.
+        """
+        # Avoid caching for any ops that would require a more sophisticated
+        # caching implementation, e.g., data dependent ops or ops that modify
+        # the inputs.
+        if torch.Tag.data_dependent_output in func.tags:
+            raise _BypassDispatchCache("data dependent output")
+
+        if torch.Tag.dynamic_output_shape in func.tags:
+            raise _BypassDispatchCache("dynamic output shape")
+
+        if torch.Tag.inplace_view in func.tags:
+            raise _BypassDispatchCache("inplace view")
+
+        if func == aten._unsafe_view.default:
+            raise _BypassDispatchCache("unsafe view")
+
+        if func == aten.set_.source_Tensor:
+            raise _BypassDispatchCache("set source")
+
+        if func in self.lift_fns:
+            raise _BypassDispatchCache("lift")
+
+        # In order to handle storage aliasing, we need to establish the alias
+        # for any view op on a cache hit. But CompositeImplicitAutograd ops may
+        # or may not alias the input, so just punt on caching these.
+        if func.is_view and torch._C._dispatch_has_kernel_for_dispatch_key(
+            func.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+        ):
+            raise _BypassDispatchCache("CompositeImplicitAutograd")
+
+        shape_env = self.shape_env
+        key_values = (
+            func,
+            # Translate any FakeTensor args to metadata.
+            self._translate_arg(args),
+            self._translate_arg(kwargs),
+            # Capture the default_dtype mode since that can affect the output tensor,
+            # e.g., when operating on constant float values.
+            torch.get_default_dtype(),
+            # Capture the current device to support, e.g., cache tensor creation,
+            # where there isn't necessarily a tensor to take the device from.
+            torch._C._get_default_device(),
+            # Shape env settings could affect behavior. One example seen in the wild:
+            # The disallowance of dynamic shapes could introduce a
+            # DynamicOutputShapeException where it wasn't seen on a previous instance
+            # of the same op.
+            shape_env.allow_scalar_outputs if shape_env else None,
+            shape_env.allow_dynamic_output_shape_ops if shape_env else None,
+            shape_env.assume_static_by_default if shape_env else None,
+            shape_env.specialize_zero_one if shape_env else None,
+            shape_env.duck_shape if shape_env else None,
```

Comment:
Can we factor this out to a dataclass like `ShapeEnvSettings` and add `xxx: add new settings to ShapeEnvSettings` comment in shape env `__init__` ? 

If we were to be extra cautious we could even commit a test which checks that all `boolean` fields in shape_env are tracked in the `ShapeEnvSettings`